### PR TITLE
New Command- Convert-PnPFileToPdf

### LINKED ISSUE #3422: [FEATURE] Convert a file in SharePoint to PDF.
**Solution Proposal**
The solution is to convert a file in SharePoint to PDF. 
A new command as ***Convert-PnPFileToPDF***
The command should accept the file url (mandatory) and the target location (as optional) where the file should be saved.
If target location isn't specified, the file will be saved in one drive.

**Will be working on this, and creating a PR soon.**

*Save to Local 
*Retrieve as Memory Stream
*Upload to SharePoint Library 
*CheckInType
*Publish
*Approve

### PR DESCRIPTION
**Pull request by NishkalankBezawada**
## Type ##
- [ ] Bug Fix
- [x] New Feature
- [ ] Sample

## Related Issues? ##
Fixes #3422

## What is in this Pull Request ? ##
New command which converts a file in SharePoint to PDF.

## Convert-PnPFileToPDF
Converts a file to Pdf, File contents can either be saved to local, returned as memorystream and can upload the file to SharePoint Online (Same Site Collection)

### Save to local path
```powershell
Convert-PnPFileToPDF -Url <String> -Path <String> [-Force]
```

### Return as memorystream
```powershell
Convert-PnPFileToPDF -Url <String> -AsMemoryStream
```

### Save to SharePoint Online (Same SiteCollection)
```powershell
Convert-PnPFileToPDF -Url <String> -Folder<String>
```

## DESCRIPTION
Allows the conversion of a file from SharePoint Online. The file contents can either be directly saved to local disk, or stored in memory for further processing, or Can be uploaded back to SharePoint Online SiteCollection

### EXAMPLE 1
```powershell
Convert-PnPFileToPDF -Url "/sites/project/Shared Documents/Document.docx" -AsMemoryStream
```

Retrieves the file, converts to PDF, and outputs its content to the console as a Memory Stream

### EXAMPLE 2
```powershell
Convert-PnPFileToPDF -Url "/sites/project/Shared Documents/Document.docx"
```

Retrieves the file, converts to PDF, and outputs its content to the console as a Memory Stream

### EXAMPLE 3
```powershell
Convert-PnPFileToPDF -Url "/sites/project/Shared Documents/Document.docx" -Path "C:\Temp"
```

Retrieves the file and converts to PDF, and save it to the given local path

### EXAMPLE 4
```powershell
Convert-PnPFileToPDF -Url "/sites/project/Shared Documents/Document.docx" -Path "C:\Temp" -Force
```

Retrieves the file and converts to PDF, and save it to the given local path. Force parameter will override the existing file in the location where the document gets saved.

### EXAMPLE 5
```powershell
Convert-PnPFileToPDF -Url "/sites/SampleTeamSite/Shared Documents/Nishkalank's/Book.xlsx" -Folder "Archive"
```

Retrieves the file and converts to PDF, and save it to the given Document library in SharePoint Online (Same Site Collection). Returns the saved file information in the console.

### -Url
The URL (server or site relative) to the file

```yaml
Type: String
Parameter Sets: (All)
Aliases: ServerRelativeUrl, SiteRelativeUrl

Required: True
Position: 0
Default value: None
Accept pipeline input: True (ByValue)
Accept wildcard characters: False
```

### -AsMemoryStream

```yaml
Type: SwitchParameter
Parameter Sets: Return as memorystream

Required: True
Position: Named
Default value: None
Accept pipeline input: False
Accept wildcard characters: False
```

### -Path
Local path where the file should be saved

```yaml
Type: String
Parameter Sets: Save to local path

Required: False
Position: Named
Default value: None
Accept pipeline input: False
Accept wildcard characters: False
```

### -Force
Overwrites the file if it exists.

```yaml
Type: SwitchParameter
Parameter Sets: Save to local path

Required: False
Position: Named
Default value: None
Accept pipeline input: False
Accept wildcard characters: False
```

### -Folder
The destination library in the site

```yaml
Type: FolderPipeBind
Parameter Sets: (UPLOADTOSHAREPOINT)

Required: True
Position: Named
Default value: None
Accept pipeline input: False
Accept wildcard characters: False
```

## RELATED LINKS

[Microsoft 365 Patterns and Practices](https://aka.ms/m365pnp)
[Convert files to PDF using Graph API](https://learn.microsoft.com/en-us/graph/api/driveitem-get?view=graph-rest-1.0&tabs=csharp#http-request)
I

### CODE REVIEW

**Comment by KoenZomers** [comment]
Really nice addition @NishkalankBezawada . Looks good to me. What do you think @gautamdsheth ?

File: src/Commands/Files/ConvertFileToPdf.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+using PnP.Framework.Utilities;
+using System.Management.Automation;
+using PnP.PowerShell.Commands.Base;
+using PnP.PowerShell.Commands.Utilities.REST;
+using System.IO;
+using PnP.PowerShell.Commands.Base.PipeBinds;
+
+namespace PnP.PowerShell.Commands.Files
+{
+    [Cmdlet(VerbsData.Convert, "PnPFileToPdf")]
```
Comment (gautamdsheth) [inline src/Commands/Files/ConvertFileToPdf.cs:11]: It needs to be `PnPFileToPDF`

File: src/Commands/Files/ConvertFileToPdf.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        public SwitchParameter Approve;
+
+        [Parameter(Mandatory = false, ParameterSetName = UPLOADTOSHAREPOINT)]
+        public string ApproveComment = string.Empty;
+
+        [Parameter(Mandatory = false, ParameterSetName = UPLOADTOSHAREPOINT)]
+        public SwitchParameter Publish;
+
+        [Parameter(Mandatory = false)]
+        public string PublishComment = string.Empty;
```
Comment (gautamdsheth) [inline src/Commands/Files/ConvertFileToPdf.cs:53]: We dont need these parameters. 
Lets keep it simple.

Download file locally
Upload it to SharePoint folder
Make it available as memory stream

File: src/Commands/Files/ConvertFileToPdf.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+        [Parameter(Mandatory = true, ParameterSetName = URLTOPATH)]
+        public string Path = string.Empty;
+
+        [Parameter(Mandatory = false, ParameterSetName = URLTOPATH)]
+        public SwitchParameter Force;
+
+        [Parameter(Mandatory = true, ParameterSetName = UPLOADTOSHAREPOINT)]
+        [ValidateNotNullOrEmpty]
+        public FolderPipeBind DocumentLibrary;
```
Comment (gautamdsheth) [inline src/Commands/Files/ConvertFileToPdf.cs:32]: Rename this parameter to Folder

File: src/Commands/Files/ConvertFileToPdf.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            byte[] response = RestHelper.GetByteArrayAsync(Connection.HttpClient, apiUrl, graphAccessToken).GetAwaiter().GetResult();
+            var fileToDownloadName = !string.IsNullOrEmpty(sourceFileName) ? sourceFileName : "Download";
+
+            switch (ParameterSetName)
+            {
+                case URLTOPATH:
+                    var fileOut = System.IO.Path.Combine(Path, $"{fileToDownloadName}.pdf");
+                    if (!Directory.Exists(Path))
+                    {
+                        Directory.CreateDirectory(Path);
```
Comment (gautamdsheth) [inline src/Commands/Files/ConvertFileToPdf.cs:87]: It should throw error if Folder doesn't exist.

**Review by gautamdsheth — CHANGES_REQUESTED** [review]

File: src/Commands/Files/ConvertFileToPdf.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+            var graphAccessToken = TokenHandler.GetAccessToken(this, $"https://{Connection.GraphEndPoint}/.default", Connection);
+            Url = Utilities.UrlUtilities.UrlDecode(Url.Replace("+", "%2B"));
+            var serverRelativeUrl = GetServerRelativeUrl(Url);
+            var fileObj = GetFileByServerRelativePath(serverRelativeUrl);
+            var siteId = PnPContext.Site.Id.ToString();
+            var listId = fileObj.ListId.ToString();
+            var itemId = fileObj.ListItemAllFields.Id.ToString();
+            var sourceFileName = fileObj.Name.ToString().Substring(0, fileObj.Name.ToString().LastIndexOf("."));
+            var apiUrl = $"https://{Connection.GraphEndPoint}/v1.0/sites/{siteId}/lists/{listId}/items/{itemId}/driveItem/content?format=pdf";
```
Comment (gautamdsheth) [inline src/Commands/Files/ConvertFileToPdf.cs:77]: Please create a new file or use an existing one , a utility , something like FileExtension or FileConvertUtility and move the logic there

**Review by gautamdsheth — CHANGES_REQUESTED** [review]

**Comment by gautamdsheth** [comment]
Hey @NishkalankBezawada - is it with you if I make changes here ? I actually found a different way to do this

**Comment by gautamdsheth** [comment]
@NishkalankBezawada - Actually, the way it will work is that it will support conversion of different files to different formats not just PDF, so if you haven't made the changes, don't start 😊

Sorry about this.

**Comment by NishkalankBezawada** [comment]
> @NishkalankBezawada - Actually, the way it will work is that it will support conversion of different files to different formats not just PDF, so if you haven't made the changes, don't start 😊
> 
> Sorry about this.

Hey Again @gautamdsheth, Thats true, its just the parameter we pass for the conversion. However, I have almost completed my changes though. I was sick back then, so couldnt complete it before. Will send you out the review mostly by tomorrow. 

Thanks,
Nishkalank

File: src/Commands/Files/ConvertFileToPdf.cs (hunk 1 above)
Comment (NishkalankBezawada) [reply src/Commands/Files/ConvertFileToPdf.cs]: Changes done in the latest commit

File: src/Commands/Files/ConvertFileToPdf.cs (hunk 2 above)
Comment (NishkalankBezawada) [reply src/Commands/Files/ConvertFileToPdf.cs]: Changes done in the latest commit

File: src/Commands/Files/ConvertFileToPdf.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            if (string.IsNullOrEmpty(Path))
+            {
+                Path = SessionState.Path.CurrentFileSystemLocation.Path;
+            }
+            else if (!System.IO.Path.IsPathRooted(Path))
+            {
+                Path = System.IO.Path.Combine(SessionState.Path.CurrentFileSystemLocation.Path, Path);
+            }
+
+            var graphAccessToken = TokenHandler.GetAccessToken(this, $"https://{Connection.GraphEndPoint}/.default", Connection);
```
[the remark this answers is not in the document]
Comment (NishkalankBezawada) [reply src/Commands/Files/ConvertFileToPdf.cs]: Changes done in the latest commit

File: src/Commands/Files/ConvertFileToPdf.cs (hunk 4 above)
Comment (NishkalankBezawada) [reply src/Commands/Files/ConvertFileToPdf.cs]: Changes done in the latest commit

File: src/Commands/Files/ConvertFileToPdf.cs (hunk 3 above)
Comment (NishkalankBezawada) [reply src/Commands/Files/ConvertFileToPdf.cs]: Changes done in the latest commit

**Comment by NishkalankBezawada** [comment]
Hello @KoenZomers & @gautamdsheth, Any update on this PR. 

Thanks,
Nishkalank

**Comment by gautamdsheth** [comment]
hi @NishkalankBezawada , sorry will have to close this PR.
It will be a part of this one , https://github.com/pnp/powershell/pull/3616 .

The mentioned PR takes care of quite a lot of different formats including PDF.

Will be crediting you for sure.

**Comment by NishkalankBezawada** [comment]
> hi @NishkalankBezawada , sorry will have to close this PR. It will be a part of this one , #3616 .
> 
> The mentioned PR takes care of quite a lot of different formats including PDF.
> 
> Will be crediting you for sure.

That was my idea, and almost stole it @gautamdsheth.

I would have expected a help or may be you might have modified it later by enhancing it. Not totally satisified @gautamdsheth @KoenZomers 

//Nishkalank

**Comment by gautamdsheth** [comment]
hi @NishkalankBezawada , sorry about this. 
Totally my bad, it is not my intention at all to override your code.

How do I fix this ? The code right now , as it is, looks good and we can merge it. 
I can improve it later on.

Or , we can work together via review comments and improve it ? 
Sorry again about this 🥹🙏

**Comment by NishkalankBezawada** [comment]
> as it i

Hello @gautamdsheth 

I didnt mean that you have overwritten the code. Its just the strategy, I got this idea first to use GraphAPI, and it should be welcomed. However, You later you got another idea, to get file and convert it to desired format using PnP, which is again brilliant and can be enhanced on top of the actually PR. I do understand your time and effort but would be really nice if you have discussed this without closing this PR. 

However, thanks once again for re-opening, you may merge it and probably enhance it later with your idea on saving a file in different formats. 

Thanks,
Nishkalank Bezawada

I meant to say that, the scope for improvement should be seen.

**Comment by gautamdsheth** [comment]
Hi @NishkalankBezawada !

You are correct. 
This was not the correct way to go about with this PR, I should have at least had a conversation before.

My mistake totally , will discuss first before doing anything.

Regards,
Gautam

**Pull request merged**